### ISSUE TITLE
Multiple programs attached to same Kernel Tracepoint interfere each other

### ISSUE THREAD

**Issue by palmtenor** [iovisor/bcc#1334]
To reproduce, run two `examples/cpp/RandomRead` simultaneously:
- The second run would fail on `ioctl(PERF_EVENT_IOC_SET_BPF)`. This is similar to #872, but we won't be able to create multiple Tracing Event for the same Kernel Tracepoint, and using instances won't help here as well. 
- Moreover, after the failed attempt of the second run, the first run would also stop to output result. Probably clean-up on error handling in the second run interfered with the existing program.

As we have more 24 * 7 BPF / BCC production tools and we push people to use Kernel Tracepoint instead of kprobe, this would probably be a more serious issue now than before.

**Comment by yonghong-song** [comment]
This is a kernel limitation.

For tracepoint, you only get one "id". Hence, there can be only one perf event attached to it. By design, only one bpf program can attached to one perf event.

The following is a possible scheme:
   . permit each tracepoint to attach more than one bpf programs
   . add REMOVE_BPF_PROG interface so each prog can be removed independently.
Since there is only "id", perf event attr config will be the same. Individual open/close with the same "id" may interfere with each other. Need further study.

cc @4ast

**Comment by palmtenor** [comment]
Correct, I think the "support multiple program per Tracepoint" part probably requires some Kernel side redesign. But regardless of if we work on that now, we should probably figure out for the second attach doesn't break the first one:)

**Comment by yonghong-song** [comment]
Did some investigation on this issue and found the root cause.

In the kernel, each tracepoint has a separate data structure tp_event. When you open multiple perf_events, all of them have links to the same tp_event. That is why when you tried to attach the second BPF program to the same tracepoint, it complains.

When the second perf_event gets closed due to the attachment error, the bpf program is freed as well since they share the same tp_event. The fix is in the kernel trying not to free bpf program if not the exact perf_event which installs the bpf program.

The fix can be extended to allow multiple bpf programs attached to the same tp_event.
Basically, tp_event will have a list of (prog, install_perf_event). The item (prog, install_perf_event) will be removed if install_perf_event is to be freed. When the tracepoint is fired. All programs will be executed sequentially.

Will work on a formal patch for this. cc @4ast

**Comment by myllynen** [comment]
Has there been any progress with this since September? I'm mainly interested in this from PCP (https://pcp.io) point of view, there's been some initial discussion whether we could add support for kernel tracepoints using e.g. the PCP BCC PMDA plugin and having this limitation removed would probably be helpful with that effort (the plugin could run in the background and not prevent other tools using the same tracepoints). Thanks.

**Comment by yonghong-song** [comment]
The commit is
```
commit e87c6bc3852b981e71c757be20771546ce9f76f3
Author: Yonghong Song <yhs@fb.com>
Date:   Mon Oct 23 23:53:08 2017 -0700

    bpf: permit multiple bpf attachments for a single perf event
    
    This patch enables multiple bpf attachments for a
    kprobe/uprobe/tracepoint single trace event.
    Each trace_event keeps a list of attached perf events.
    When an event happens, all attached bpf programs will
    be executed based on the order of attachment.
```

Not sure about 4.15, but 4.16 should have it, could you give a try?

**Comment by myllynen** [comment]
Thanks for the pointer, I've now tested this on 4.16 and 4.17 and it seems to partially work.

Using the urandomread.py example the second invocation fails on 4.13 (as expected) but works on 4.16/4.17. But there's still a problem. With one instance of urandomread.py running and mktemp(1) being run three times on a different terminal I see this which is fine (though not sure why that `<...>` is being printed occasionally):

```
1321.609039000     <...>            7347    80
1322.379914000     mktemp           7348    80
1323.092013000     mktemp           7349    80
```

However, with two instances of urandomread.py running and mktemp(1) invoked three times I see something like this:

```
1375.131577000     mktemp           7360    80
1375.131588000     mktemp           7360    80
1375.874894000     mktemp           7361    80
1375.874906000     mktemp           7361    80
1376.944264000     <...>            7362    80
1376.944275000     <...>            7362    80
```

So the other urandomread.py is reporting things twice and the second is not reporting anything. Sometimes it could be that the other is reporting two and the other four events, looks like they end up randomly to the clients. Do you think this is something that could be fixed?

**Comment by myllynen** [comment]
Do we know anything more about this, does this look like a kernel or a bcc issue? Thanks.

**Comment by myllynen** [comment]
Hmm, care to elaborate why this was closed, the issue is still present on Fedora 28 with kernel 4.17 / bcc 0.6.0? Thanks.

**Comment by yonghong-song** [comment]
@myllynen we do not do branching to fix old problems as this point. If necessary, you can use the trunk or you can checkout release 0.6.0/0.6.1 and do a local backport for this problem. Thanks!
